test/io/test_tim_grabber.cpp

```diff
@@ -0,0 +1,141 @@
+  public: using TimGrabber::updateLookupTables;
+  public: using TimGrabber::toPointClouds;
+};
+
+class TimGrabberTest : public ::testing::Test {
+  protected:
+    TimGrabberTest () = default;
+    ~TimGrabberTest () override {}
+
+    static void SetUpTestCase () {
+      packets_ = new std::vector<std::string> ();
+      correct_clouds_ = new std::vector<CloudT> ();
+      grabber_ = new TestableTimGrabber ();
+    }
+
+    static void TearDownTestCase () {
+      delete correct_clouds_;
+      correct_clouds_ = nullptr;
+      delete packets_;
+      packets_ = nullptr;
+      delete grabber_;
+      grabber_ = nullptr;
+    }
+
+    virtual void SetUp () {
+      constexpr float angle_start = - 1.0 * M_PI / 4.0;
+      constexpr float angle_range = 2.0 * M_PI * 3.0 / 4.0;
+
+      std::default_random_engine generator;
+      std::uniform_int_distribution<int> i_distribution (0, 1000);
+      std::uniform_real_distribution<float> f_distribution (0.0, 20.0);
+
+      CloudT cloud;
+
+      for (int i=0; i<1000; ++i) {
+        const size_t amount_of_data = i_distribution (generator);
+
+        cloud.reserve (amount_of_data);
+        cloud.clear ();
+
+        std::ostringstream ss;
+        ss << std::hex;
+        ss << " " << amount_of_data;
+
+        float angle = angle_start;
+        const float angle_step = angle_range / amount_of_data;
+
+        for (size_t i=0; i<amount_of_data; ++i, angle += angle_step) {
+          float distance = f_distribution (generator);
+          cloud.emplace_back (distance * std::cos (angle), distance * std::sin (angle), 0.0);
+          ss << " " << static_cast<int>(distance * 1000);
+        }
+        correct_clouds_->push_back (cloud);
+
+        std::string header_sample = "sRA LMDscandata 1 1 1291B11 0 0 AED5 AED7 FDB36397 FDB3779F 0 0 1 0 0 5DC A2 0 1 DIST1 3F800000 00000000 FFF92230 D05";
+        std::string packet = header_sample + ss.str ();
+        packets_->push_back (packet);
+      }
+    }
+
+    virtual void TearDown () {
+    }
+
+    static std::vector<std::string>* packets_;
+    static std::vector<CloudT>* correct_clouds_;
+    static TestableTimGrabber* grabber_;
+};
+
+std::vector<std::string>* TimGrabberTest::packets_ = nullptr;
+std::vector<CloudT>* TimGrabberTest::correct_clouds_ = nullptr;
+TestableTimGrabber* TimGrabberTest::grabber_ = nullptr;
+
+TEST_F (TimGrabberTest, Test1)
+{
+  CloudT::ConstPtr answer_cloud = grabber_->point_cloud_xyz_ptr_;
+
+  for (std::size_t i=0; i<packets_->size (); ++i) {
+    std::string packet = packets_->at(i);
+
+    grabber_->processTimPacket (packet);
+    grabber_->updateLookupTables ();
+    grabber_->toPointClouds ();
+
+    ASSERT_EQ (correct_clouds_->at(i).size (), answer_cloud->size ());
+
+    for (std::size_t j = 0; j < correct_clouds_->at(i).size (); j++)
+    {
+      const PointT &correct_point = correct_clouds_->at(i).at(j);
+      const PointT &answer_point = answer_cloud->at(j);
+      if (std::isnan (correct_point.x))
```

Comment:
Insert Nan in the data in order to actually test the nan checks

Reply:
I wrote this check (`isnan()`) referenced by `test/io/test_grabbers.cpp`.

But  I think about it again, these comparison (some `isnan()`) are not necessary.
Because in this test, `0<= x,y,z <= 20` was generated by random_generator.

Required comparison are only
- correct.x == answer.x
- correct.y == answer.y
- correct.z == answer.z

Reply:
Is there any situation in which nan will be in the input from the real sensor? If no, then all these checks are not needed in the grabber (and the test)

Reply:
No!

I used Tim to send light to the night sky now,  but there was no NAN value.

※ The maximum distance of Tim is 20m.